src/mmds/src/token_headers.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use micro_http::{HttpHeaderError, RequestError};
+use std::collections::HashMap;
+use std::fmt;
+use std::fmt::{Display, Formatter};
+use std::result::Result;
+
+/// Wrapper over supported HTTP Custom Header types.
+#[derive(Debug, Eq, Hash, PartialEq)]
+pub enum TokenHeader {
+    /// Header `X-metadata-token`
+    XMetadataToken,
+    /// Header `X-metadata-token-ttl-seconds`
+    XMetadataTokenTtlSeconds,
+}
+
+impl Display for TokenHeader {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match *self {
+            TokenHeader::XMetadataToken => write!(f, "X-metadata-token"),
+            TokenHeader::XMetadataTokenTtlSeconds => {
+                write!(f, "X-metadata-token-ttl-seconds")
+            }
+        }
+    }
+}
+
+/// Wrapper over the list of token headers associated with a Request.
+#[derive(Debug, PartialEq)]
+pub struct TokenHeaders {
+    /// The `X-metadata-token` header might be used by HTTP clients to specify a token in order
+    /// to authenticate to the session. This is used for guest requests to MMDS only.
+    x_metadata_token: Option<String>,
+    /// The `X-metadata-token-ttl-seconds` header might be used by HTTP clients to specify
+    /// the expiry time of a token. This is used for PUT requests issued by the guest to MMDS only.
+    x_metadata_token_ttl_seconds: Option<u32>,
+}
+
+impl Default for TokenHeaders {
+    /// Token headers are not present in the request by default.
+    fn default() -> Self {
+        Self {
+            x_metadata_token: None,
+            x_metadata_token_ttl_seconds: None,
+        }
+    }
+}
+
+impl TokenHeaders {
+    /// Return `TokenHeaders` from headers map.
+    pub fn try_from(map: &HashMap<String, String>) -> Result<TokenHeaders, RequestError> {
+        let mut headers = Self::default();
+
+        if let Some(token) = map.get(&*TokenHeader::XMetadataToken.to_string()) {
+            headers.x_metadata_token = Some(token.to_string());
+        }
+
+        if let Some(value) = map.get(&*TokenHeader::XMetadataTokenTtlSeconds.to_string()) {
+            match value.parse::<u32>() {
+                Ok(seconds) => {
+                    headers.x_metadata_token_ttl_seconds = Some(seconds);
```

Comment:
Should we validate if `seconds <= MAX_TTL` ?

Reply:
We do validate this as part of the token generation process.